gen3utils/etl/etl_validator.py

```diff
@@ -1,202 +1,275 @@
+                for prop in v.get("props"):
+                    new_props = validate_prop(
+                        prop,
+                        labels_to_back_refs.values(),
+                        nodes_with_props,
+                        recorded_errors,
+                        labels_to_back_refs.get(k),
+                    )
+                    index.props.update({p.name: p for p in new_props})
+
+
+def validate_joining_prop(json_obj, recorded_errors, joining_index):
+    name = validate_name(json_obj, recorded_errors)
+    validate_fn(json_obj, recorded_errors)
+    validate_joining_src(json_obj, recorded_errors, joining_index.props)
+    return Prop(name)
+
+
+def validate_prop(
+    json_obj, list_of_nodes, nodes_with_props, recorded_errors, grouping_path=None
+):
+    names = validate_path(
+        json_obj, grouping_path, recorded_errors, list_of_nodes, nodes_with_props
+    )
+    if len(names) == 0:
+        names.append(
+            validate_name_src(
+                json_obj,
+                json_obj.get("path", grouping_path),
+                recorded_errors,
+                nodes_with_props,
+            )
+        )
+
+    props = [Prop(n) for n in names]
+    return props
+
+
+def validate_joining_src(json_obj, recorded_errors, joining_props):
+    src = json_obj.get("src", json_obj.get("name"))
+    if src is not None:
+        if src not in joining_props:
+            recorded_errors.append(
+                FieldError(
+                    "src field {} (declared in {}) is not found in joining index.".format(
+                        src, json_obj
+                    )
+                )
+            )
+    else:
+        recorded_errors.append(
+            FieldError("Missing source field for {}".format(json_obj))
+        )
+
+
+def validate_fn(json_obj, recorded_errors):
+    fn = json_obj.get("fn")
+    if fn is not None:
+        if fn not in ["set", "count", "list", "sum", "min", "max"]:
+            recorded_errors.append(
+                MappingError(
+                    "{} function (declared in {}) is not supported in ETL".format(
+                        fn, json_obj
+                    ),
+                    "Function",
+                )
+            )
+    return fn
+
+
+def validate_name(json_obj, recorded_errors):
+    name = json_obj.get("name")
+    if name is None:
+        recorded_errors.append(
+            PropertiesError("Missing name for mapping property {}.".format(json_obj))
+        )
+    elif name == "":
+        recorded_errors.append(
+            PropertiesError(
+                "Name should not be empty for mapping property {}.".format(json_obj)
+            )
+        )
+    return name
+
+
+def validate_name_src(json_obj, path, recorded_errors, nodes_with_props):
+    name = validate_name(json_obj, recorded_errors)
+    fn = validate_fn(json_obj, recorded_errors)
+    src = json_obj.get("src", name)
+    if src is not None:
```

Comment:
is it ok if src is None? there is no else clause

if it's ok, you can just do this and remove one level of indentation
```
if not src:
    return name

if not path:
    ...
```